File: leetcode-js/256-paint-house.js
/**
 * @param {number[][]} costs
 * @return {number}
 */
var minCost = function(costs) {
    if (!costs.length) {
        return 0;
    }

    const dp = [costs[0]];

    for (let i = 1; i < costs.length; i += 1) {
        dp[i] = [
            costs[i][0] + Math.min(dp[i-1][1], dp[i-1][2]),
            costs[i][1] + Math.min(dp[i-1][0], dp[i-1][2]),
            costs[i][2] + Math.min(dp[i-1][0], dp[i-1][1])
        ];
    }

    return Math.min(
        dp[costs.length-1][0],
        dp[costs.length-1][1],
        dp[costs.length-1][2]
    );
};

File: leetcode-java/256-paint-house.java
public class Solution {
    public int minCost(int[][] costs) {
        int n = costs.length;
        int r = 0;
        int g = 0;
        int b = 0;
        for (int i = 0; i < n; i++) {
            int rr = r;
            int gg = g;
            int bb = b;
            r = costs[i][0] + Math.min(bb, gg);
            g = costs[i][1] + Math.min(rr, bb);
            b = costs[i][2] + Math.min(rr, gg);
        }
        return Math.min(Math.min(r, g), b);
    }
}
